src/StudyEntities/Candidate/IdentifierGenerator.php

```diff
@@ -0,0 +1,138 @@
+<?php declare(strict_types=1);
+
+namespace LORIS\StudyEntities\Candidate;
+
+abstract class IdentifierGenerator
+{
+    protected $generationMethod;
+    protected $alphabet = array();
+    protected $length;
+    protected $minValue;
+    protected $maxValue;
+    protected $prefix    = '';
+
+    /**
+    * Generates a new unused identifier to represent a Candidate in LORIS.
+    *
+    * @return string The new identifier
+    */
+    public function createNewID(): string
+    {
+        $this->checkIDRangeFull();
+
+        $id = '';
+        switch ($this->generationMethod) {
+        case 'sequential':
+            $id = $this->padID($this->generateSequentialID());
+            break;
+        case 'random':
+            $id = $this->padID($this->generateRandomID());
+            // If the Random ID is already in use or is outside of the range 
+            // defined by minValue and maxValue, increment it by using the
+            // generateSequentialID function with the previous ID as a starting
+            // point.
+            if (in_array($id, $this->getExistingIDs(), true)
+                || strcmp(strval($id), strval($this->minValue)) < 0
+                || strcmp(strval($id), strval($this->maxValue)) > 0
+            ) {
+                $id = $this->generateSequentialID($id);
+            }
+            break;
+        }
+
+        // Return padded value with appropriate prefix
+        return $this->prefix . $id;
+    }
+
+    protected function generateSequentialID(string $id = ''): string
+    {
+        $newID = $id ?? max($this->getExistingIDs());
```

Comment:
`$id` will always equal the empty string if not given an other value. The max($this->getExistingIDs()) will never be evaluated